Firestore/core/src/core/expressions_eval.cc

```diff
@@ -0,0 +1,182 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/core/expressions_eval.h"
+
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/api/expressions.h"
+#include "Firestore/core/src/api/stages.h"
+#include "Firestore/core/src/model/mutable_document.h"
+#include "Firestore/core/src/remote/serializer.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+EvaluateResult::EvaluateResult(
+    EvaluateResult::ResultType type,
+    nanopb::Message<google_firestore_v1_Value> message)
+    : value_(std::move(message)), type_(type) {
+}
+
+EvaluateResult EvaluateResult::NewNull() {
+  return EvaluateResult(
+      ResultType::kNull,
+      nanopb::Message<google_firestore_v1_Value>(model::MinValue()));
+}
+
+EvaluateResult EvaluateResult::NewValue(
+    nanopb::Message<google_firestore_v1_Value> value) {
+  if (model::IsNullValue(*value)) {
+    return EvaluateResult::NewNull();
+  } else if (value->which_value_type ==
+             google_firestore_v1_Value_boolean_value_tag) {
+    return EvaluateResult(ResultType::kBoolean, std::move(value));
+  } else if (model::IsInteger(*value)) {
+    return EvaluateResult(ResultType::kInt, std::move(value));
+  } else if (model::IsDouble(*value)) {
+    return EvaluateResult(ResultType::kDouble, std::move(value));
+  } else if (value->which_value_type ==
+             google_firestore_v1_Value_timestamp_value_tag) {
+    return EvaluateResult(ResultType::kTimestamp, std::move(value));
+  } else if (value->which_value_type ==
+             google_firestore_v1_Value_string_value_tag) {
+    return EvaluateResult(ResultType::kString, std::move(value));
+  } else if (value->which_value_type ==
+             google_firestore_v1_Value_bytes_value_tag) {
+    return EvaluateResult(ResultType::kBytes, std::move(value));
+  } else if (value->which_value_type ==
+             google_firestore_v1_Value_reference_value_tag) {
+    return EvaluateResult(ResultType::kReference, std::move(value));
+  } else if (value->which_value_type ==
+             google_firestore_v1_Value_geo_point_value_tag) {
+    return EvaluateResult(ResultType::kGeoPoint, std::move(value));
+  } else if (model::IsArray(*value)) {
+    return EvaluateResult(ResultType::kArray, std::move(value));
+  } else if (model::IsVectorValue(*value)) {
+    // vector value must be before map value
+    return EvaluateResult(ResultType::kVector, std::move(value));
+  } else if (model::IsMap(*value)) {
+    return EvaluateResult(ResultType::kMap, std::move(value));
+  } else {
+    return EvaluateResult(ResultType::kError, {});
```

Comment:
nit: you can use brace initialization for all of `EvaluateResult(...)` here. e.g.

```cpp
    return {ResultType::kBoolean, std::move(value)};
```

Reply:
This is a intentional change to make things explicit, hopefully this is OK?